composite/market-closure/src/adapter.ts

```diff
@@ -0,0 +1,43 @@
+import { AdapterRequest, AdapterResponse, Execute } from '@chainlink/types'
+import { Requester, Validator } from '@chainlink/external-adapter'
+import { CheckExecute, getCheck, getCheckImpl } from './checks'
+import { getImpl, getPriceDataProvider } from './price'
+import { getLatestAnswer } from '@chainlink/reference-data-reader'
+
+const customParams = {
+  base: ['base', 'asset', 'from'],
+  referenceContract: ['referenceContract', 'contract'],
+  multiply: true,
+  schedule: false,
+}
+
+export const execute: Execute = async (input) => {
+  const adapterExecute = getImpl({ type: getPriceDataProvider() })
+  const checkExecute = getCheckImpl({ type: getCheck() })
+  return await executeWithAdapters(input, adapterExecute, checkExecute)
+}
+
+export const executeWithAdapters = async (
+  input: AdapterRequest,
+  adapter: Execute,
+  check: CheckExecute,
+): Promise<AdapterResponse> => {
+  const validator = new Validator(input, customParams)
+  if (validator.error) throw validator.error
+
+  const jobRunID = validator.validated.id
+  const symbol = validator.validated.data.base.toUpperCase()
+  const schedule = validator.validated.data.schedule || {}
+  const referenceContract = validator.validated.data.referenceContract
+  const multiply = validator.validated.data.multiply
+
+  const halted = await check(symbol, schedule)
+  if (halted) {
+    const result = await getLatestAnswer(referenceContract, multiply, input.meta)
+    return Requester.success(jobRunID, { data: { result }, result, status: 200 })
+  }
+
+  return await adapter(input)
+}
```

Comment:
We mix few adapters and their input prams in one place. I suggest to structure this differently:

```suggestion
export const execute: Execute = async (input) => {
  const check = getCheckImpl(getCheckProvider(), input)
  const shouldReadOnChain = await check() // no input params, already prepared fn
  if (shouldReadOnChain) return await getLatestAnswer(input)
  
  const adapter = getImpl(getPriceDataProvider())
  return await adapter(input)
}

const getLatestAnswer = async (input: AdapterRequest): Promise<AdapterResponse> => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const referenceContract = validator.validated.data.referenceContract
  const multiply = validator.validated.data.multiply

  // TODO: use @chainlink/contracts
  const result = await getLatestAnswer(referenceContract, multiply, input.meta)
  return Requester.success(jobRunID, { data: { result }, result, status: 200 })
}

```